pkg/catalog/egress.go

```diff
@@ -1,12 +1,188 @@
 	"github.com/openservicemesh/osm/pkg/identity"
+	"github.com/openservicemesh/osm/pkg/service"
 	"github.com/openservicemesh/osm/pkg/trafficpolicy"
 )
 
+const (
+	protocolHTTP = "http"
+)
+
 // GetEgressTrafficPolicy returns the Egress traffic policy associated with the given service identity
 func (mc *MeshCatalog) GetEgressTrafficPolicy(serviceIdentity identity.ServiceIdentity) (*trafficpolicy.EgressTrafficPolicy, error) {
-	// TODO(#3045): Implement egres policies
-	return nil, nil
+	var trafficMatches []*trafficpolicy.TrafficMatch
+	var clusterConfigs []*trafficpolicy.EgressClusterConfig
+	allowedDestinationPorts := mapset.NewSet()
+	portToRouteConfigMap := make(map[int][]*trafficpolicy.EgressHTTPRouteConfig)
+
+	egressResources := mc.policyController.ListEgressPoliciesForSourceIdentity(serviceIdentity.ToK8sServiceAccount())
+
+	for _, egress := range egressResources {
+		for _, portSpec := range egress.Spec.Ports {
+			// ---
+			// Build the HTTP route configs for the given Egress policy
+			if strings.EqualFold(portSpec.Protocol, protocolHTTP) {
+				httpRouteConfigs, httpClusterConfigs := mc.buildHTTPRouteConfigs(egress, portSpec.Number)
+				portToRouteConfigMap[portSpec.Number] = append(portToRouteConfigMap[portSpec.Number], httpRouteConfigs...)
+				clusterConfigs = append(clusterConfigs, httpClusterConfigs...)
+			}
+
+			// ---
+			// TODO(#3045): Build the TCP route configs for the given Egress policy
+
+			// ---
+			// Build traffic matches for the given Egress policy.
+			// Traffic matches are used to match outbound traffic as egress traffic using the port numbers
+			// specified in Egress policies.
+			newlyAdded := allowedDestinationPorts.Add(portSpec)
+			if newlyAdded {
+				trafficMatches = append(trafficMatches, &trafficpolicy.TrafficMatch{
+					DestinationPort: portSpec,
+				})
+			}
+		}
+	}
+
+	return &trafficpolicy.EgressTrafficPolicy{
+		HTTPRouteConfigsPerPort: portToRouteConfigMap,
+		TrafficMatches:          trafficMatches,
+		ClustersConfigs:         clusterConfigs,
+	}, nil
+}
+
+func (mc *MeshCatalog) buildHTTPRouteConfigs(egressPolicy *policyV1alpha1.Egress, port int) ([]*trafficpolicy.EgressHTTPRouteConfig, []*trafficpolicy.EgressClusterConfig) {
+	if egressPolicy == nil {
+		return nil, nil
+	}
+
+	var routeConfigs []*trafficpolicy.EgressHTTPRouteConfig
+	var clusterConfigs []*trafficpolicy.EgressClusterConfig
+
+	// Before building the route configs, pre-compute the allowed IP ranges since they
+	// will be the same for every HTTP route config derived from the given Egress policy.
+	var allowedDestinationIPRanges []string
+	destIPSet := mapset.NewSet()
+	for _, ipRange := range egressPolicy.Spec.IPAddresses {
+		if _, _, err := net.ParseCIDR(ipRange); err != nil {
+			log.Error().Err(err).Msgf("Invalid IP range [%s] specified in egress policy %s/%s; will be skipped", ipRange, egressPolicy.Namespace, egressPolicy.Name)
+			continue
+		}
+		newlyAdded := destIPSet.Add(ipRange)
```

Comment:
Reading `mapset`'s `.Add()` - this will return `//False if it existed already` (for my own FYI)